### DISCUSSION TITLE
Dead positions CHA can see but Stockfish not

### DISCUSSION THREAD

**Discussion by dlbbld** [official-stockfish/Stockfish#3916] [Q&A]
The surely well-known project from Ambrona can identify some positions as dead positions, Stockfish cannot yet see, like the one below where Stockfish sees Black winning:

8/1k5B/7b/8/1p1p1p1p/1PpP1P1P/2P3K1/N3b3 b - - 0 1

https://github.com/miguel-ambrona/D3-Chess
https://elrubiongamma.ddns.net/chess-unwinnability-analyzer/index.html

In this respect, I would like to learn more about this topic. Is there more information on this topic, like an assessment of this project?

**Comment by pb00067** [comment]
As I know the primary goal of Stockfish is to get as strong a possible in gameplay.
SF is not designed to be a (good) analysis tool.
Integrating such features as recognizing dead positions would not improve Stockfish's gameply, it would rather weaken it because dead position recognition is very complex.
On the other hand SF is open source so everybody interested can fork the project and create own versions/specializations of SF.
(For instance there exists a fork/branch of SF which is specialized in recognizing mate positions, unfortunately not developed anymore:
https://github.com/jhellis3/Stockfish/tree/mate_finder  )

**Reply by dlbbld** [reply]
Thanks for the comments.

I do not assume that an improved identification of dead positions improves playing strength overall. But it might improve playing strength in some special situations. For example, for the above position Q, where Stockfish sees Black as winning.

Let's assume some position P and a move m is leading to position Q, which Stockfish gives a score of 2.5 instead of 0.0. The move n leads to a position R, which is really winning but only with a score of 1.0. So in this made-up example, one would choose m as the better move, but n is the better move. Knowing that position Q is dead, one could asses n as the better move, so determine a better move for this position.

In this respect, I assume that an improved dead position detection could improve finding the best move for some positions. 

Of course, Stockfish works in another way. So the above will not apply 1:1. Still, the dead position detection by CHA has a high coverage of dead position detection (around maybe 99.9% or more, and on the other saying, it can't detect), it is worth a second glance.

**Comment by vondele** [comment]
I think the project is definitely interesting. There is usually little we can do for single positions. However, if very large numbers of positions are available (at least 10000s) one could consider to use some of the data to train the network. Do you know how large the database is of all blocked positions?

**Reply by dlbbld** [reply]
Thanks for your interest. First of all, all that I write is second-hand information. You might want to contact the author, which can give you excellent first-hand information. The contact he indicates on the website is https://lichess.org/@/ambrona.

Yes, there is a dead position list, or more precisely, a position list where at least one player or both cannot mate, ~~and the reason is not insufficient material~~ the previous statement is wrong. The list also contains one side dead positions due to insufficient material (e.g. bare king). About one-third of the Lichess database was analyzed as part of the project.

It resulted in about 100'000 dead positions. If analyzed, I guess that the other two-thirds contain more dead positions, as players in dead positions often draw.

I quote the author below:
https://chasolver.org/
Experimental Results
> We have evaluated CHA over the entire [Lichess Open Database](https://database.lichess.org/) of standard rated games, which includes 3,546,880,334 games to date (July 2022). More concretely, we have applied CHA to the final position of all games that ended in a timeout and that were classified as 1-0 or 0-1. This represents a total of 1,125,259,435 games (about 32% of all games) which have been analyzed in about 111 hours of CPU time (354 μs per position on average).
> 
> Our analysis led to identifying a total of [94,858 games](https://chasolver.org/games.html) that were unfairly classified. Namely, games that were lost by the player who ran out of time, but their opponent could not have checkmated them by any possible sequence of legal moves.

**Comment by dlbbld** [comment]
Just for convenience, the list of the positions is in the quote. The entries with "blocked" are dead positions due to the position, not material. I assume these are the ones of interest:
https://chasolver.org/games.html
